### PR TITLE
Change substring() to use `startIndex, endIndex` instead of `startIndex, length`

### PR DESCRIPTION
**Pull request by ryn5**

### CODE REVIEW

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -25,6 +25,7 @@ image::https://raw.githubusercontent.com/apache/tinkerpop/master/docs/static/ima
 
 This release also includes changes from <<release-3-6-5, 3.6.6>> and <<release-3-5-8, 3.5.8>>.
 
+* Changed `substring()` to use `startIndex, endIndex` instead of `startIndex, length`."
```
Comment (xiazcy) [inline CHANGELOG.asciidoc:28]: We don't need to put this since it's not released yet, just update the documentation for it.

File: docs/src/upgrade/release-3.7.x.asciidoc [hunk 2]
```diff
@@ -206,11 +206,12 @@ String index out of range: -2
 Type ':help' or ':h' for help.
 ----
 
-The `substring()`-step will begin at the start index and return a substring with the length specified. Negative start
-indices are allowed and will begin at the specified index counted from the end of the string:
+The `substring()`-step will return a substring with indices specified by the start and end indices, or from
+the start index to the remainder of the string if an end index is not specified. Negative indices are allowed and will
+begin at the specified index counted from the end of the string:
```
Comment (Cole-Greer) [inline docs/src/upgrade/release-3.7.x.asciidoc:211]:
```suggestion
the start index to the remainder of the string if an end index is not specified. Negative indices are allowed and will
count from the end of the string:
```
Suggest removing "begin" as that specifically refers to the start index

**Review by Cole-Greer — APPROVED** [review]
Looks good to me, just a few minor docs nits and I think it's ready for an external PR.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java [hunk 3]
```diff
@@ -104,4 +104,15 @@ private int processStartIndex(int strLen) {
         }
     }
 
+    // Helper function to process the end index, if it is negative (which counts from end of string) it is converted
+    // to the positive index position or 0 when negative index exceeds the string length, which means an empty string
+    // will be returned, if it is positive and exceeds the length of the string, it is converted to the length of the
+    // string.
+    private int processEndIndex(int strLen) {
```
Comment (kenhuuu) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java:111]: This is basically the same function as processStartIndex(). Consider refactoring.

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStepTest.java [hunk 4]
```diff
@@ -41,16 +41,26 @@ protected List<Traversal> getTraversals() {
...
+        assertEquals("", __.__("hello world").substring(1, 0).next());
+        assertEquals("hello worl", __.__("hello world").substring(0, -1).next());
+        assertEquals("", __.__("hello world").substring(-1, 0).next());
+        assertEquals("ello worl", __.__("hello world").substring(1, -1).next());
+        assertEquals("", __.__("hello world").substring(-1, 1).next());
+        assertEquals("d", __.__("hello world").substring(-1, 11).next());
+        assertEquals("ello world", __.__("hello world").substring(1, 11).next());
+        assertEquals("d", __.__("hello world").substring(10, 11).next());
+        assertEquals("hello world", __.__("hello world").substring(-11, 11).next());
+        assertEquals("h", __.__("hello world").substring(-11, 1).next());
```
Comment (kenhuuu) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStepTest.java:61]: Missing a test for two negative numbers like (-3,-2).

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Substring.feature [hunk 5]
```diff
@@ -55,11 +55,11 @@ Feature: Step - substring()
       | p |
       | pple |
```
Comment (xiazcy) [inline gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Substring.feature:57]: General comment, I would add some feature test cases with combinations of (pos start index, neg end index), (neg start index, pos end index), and (neg start index, neg end index).

File: CHANGELOG.asciidoc (hunk 1 above)
Comment (ryn5) [reply CHANGELOG.asciidoc]: Removed line

File: docs/src/upgrade/release-3.7.x.asciidoc (hunk 2 above)
Comment (ryn5) [reply docs/src/upgrade/release-3.7.x.asciidoc]: Committed suggestion

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java [hunk 6]
```diff
@@ -1608,20 +1608,20 @@ public default GraphTraversal<S, String> substring(final int startIndex) {
     }
 
     /**
-     * Returns a substring of the incoming string traverser with a 0-based start index (inclusive) index and length
-     * specified. If the start index is negative then it will begin at the specified index counted from the end
-     * of the string, or 0 if exceeding the string length. Length is optional, if it is not specific or if it exceeds
-     * the length of the string then all remaining characters will be returned. Length <= 0 will return the empty string.
-     * Null values are not processed and remain as null when returned. If the incoming traverser is a non-String value then an
-     * {@code IllegalArgumentException} will be thrown.
+     * Returns a substring of the incoming string traverser with a 0-based start index (inclusive) index and end index
```
[the remark this answers is not in the document]
Comment (ryn5) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java]: Committed suggestion

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java (hunk 3 above)
Comment (ryn5) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java]: Merged into processStringIndex()

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStepTest.java (hunk 4 above)
Comment (ryn5) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStepTest.java:61]: Added (-3, -1) and (-1, -3) cases

File: docs/src/dev/provider/gremlin-semantics.asciidoc [hunk 7]
```diff
@@ -1194,7 +1194,7 @@ link:https://tinkerpop.apache.org/docs/x.y.z/reference/#split-step[reference]
 [[substring-step]]
 === substring()
 
-*Description:* Returns a substring of the incoming string traverser with a 0-based start index (inclusive) and length specified.
+*Description:* Returns a substring of the incoming string traverser with a 0-based start index (inclusive) and end index (exclusive).
 
 *Syntax:* `substring(long, long)`
```
[the remark this answers is not in the document]
Comment (ryn5) [reply docs/src/dev/provider/gremlin-semantics.asciidoc:1199]: Fixed
